tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,753 @@
+		cdm.curFileMint = mint
+		// As this wont happen a whole lot of time, we don't wait
+		// for a new head chunk file to be cut to write it off.
+		binary.BigEndian.PutUint64(cdm.writeBuf[:], uint64(mint))
+		if _, err := cdm.curFile.WriteAt(cdm.writeBuf[:MintMaxtSize], HeaderMintOffset); err != nil {
+			return 0, err
+		}
+	}
+
+	cdm.inBufferChunksLocks[chunkRef%inBufferShards].Lock()
+	cdm.inBufferChunks[chunkRef%inBufferShards][chunkRef] = chk
+	cdm.inBufferChunksLocks[chunkRef%inBufferShards].Unlock()
+
+	if len(chk.Bytes()) >= writeBufferSize-30 {
+		if err := cdm.flushBuffer(); err != nil {
+			return 0, err
+		}
+	}
+
+	return chunkRef, nil
+}
+
+func (cdm *ChunkDiskMapper) chunkRef(seq, offset uint64) (chunkRef uint64) {
+	return (seq << 32) | offset
+}
+
+func (cdm *ChunkDiskMapper) shouldCutNewFile(chunkLength int, maxt int64) bool {
+	return cdm.n == 0 || // First head chunk file.
+		(maxt-cdm.curFileMint > cdm.maxFileMs && cdm.n > HeadChunkFileHeaderSize) || // Time duration reached for the existing file.
+		cdm.n+int64(chunkLength+MaxHeadChunkMetaSize) >= MaxHeadChunkFileSize // Crosses the max head chunk file size.
+}
+
+func (cdm *ChunkDiskMapper) cut(mint int64) (returnErr error) {
+	// Sync current tail to disk and close.
+	if err := cdm.finalizeCurFile(); err != nil {
+		return err
+	}
+
+	n, f, seq, err := cutSegmentFile(cdm.dir, MagicHeadChunks, headChunksFormatV1, 0)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		// The file should not be closed if there is no error,
+		// its kept open in the ChunkDiskMapper.
+		if returnErr != nil {
+			var merr tsdb_errors.MultiError
+			merr.Add(returnErr)
+			merr.Add(f.Close())
+			returnErr = merr.Err()
+		}
+	}()
+
+	cdm.size += cdm.n
+	atomic.StoreInt64(&cdm.n, int64(n))
+	oldSeq := cdm.curFileSequence
+	cdm.curFileSequence = seq
+
+	// Write the provided time for mint and 0 for maxt.
+	cdm.curFileMint = mint
+	binary.BigEndian.PutUint64(cdm.writeBuf[:], uint64(mint))
+	if _, err := f.Write(cdm.writeBuf[:MintMaxtSize]); err != nil {
+		return err
+	}
+	binary.BigEndian.PutUint64(cdm.writeBuf[:], 0)
+	if _, err := f.Write(cdm.writeBuf[:MintMaxtSize]); err != nil {
+		return err
+	}
+	atomic.AddInt64(&cdm.n, 16)
+
+	oldFile := cdm.curFile
+
+	cdm.curFile = f
+	if cdm.wbuf != nil {
+		cdm.wbuf.Reset(f)
+	} else {
+		cdm.wbuf = bufio.NewWriterSize(f, writeBufferSize)
+	}
+
+	if oldFile != nil {
+		// Open it again with the new size.
+		newTailFile, err := fileutil.OpenMmapFile(oldFile.Name())
+		if err != nil {
+			return err
+		}
+		cdm.mmappedChunksMtx.Lock()
+		oldMmapFile := cdm.closers[oldSeq]
+		cdm.closers[oldSeq] = newTailFile
+		cdm.mmappedChunkFiles[oldSeq] = realByteSlice(newTailFile.Bytes())
+		cdm.mmappedChunksMtx.Unlock()
+		// Closing the last mmapped file.
+		if err := oldMmapFile.Close(); err != nil {
+			cdm.mmappedChunksMtx.Unlock()
+			return err
+		}
+	}
+
+	mmapFile, err := fileutil.OpenMmapFileWithSize(f.Name(), int(MaxHeadChunkFileSize))
```

Comment:
If we want those locks (probably not so many), we might want to have some helper for this. (:

Reply:
I did not understand what you meant here :)

Reply:
I mean that we have tons of locks and proposing a helper, but I think you reduce a bit, so will review.